File: test/corpus/propagation/call_argument.exs
defmodule RustQ.Corpus.Propagation.CallArgument do
  @moduledoc "Argument-position propagation inferred from local callable metadata."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec decode(atom()) :: R.nif_result(R.u32())
  defrust decode(atom) do
    decode_as!(atom, R.u32())
  end

  @spec consume(R.u32()) :: R.nif_result(R.unit())
  defrust consume(value) do
    _copy = value
    :ok
  end

  @spec run(atom()) :: R.nif_result(R.unit())
  defrust run(atom) do
    consume(decode(atom))
    :ok
  end
end

File: test/corpus/propagation/call_argument.rs
fn decode(atom: Atom) -> NifResult<u32> {
    atom.decode::<u32>()?
}

fn consume(value: u32) -> NifResult<()> {
    let _copy = value;
    Ok(())
}

fn run(atom: Atom) -> NifResult<()> {
    consume(decode(atom)?)?;
    Ok(())
}
